src/plugins/ScryptEncrypter.ts

```diff
@@ -0,0 +1,145 @@
+import scrypt from "scrypt-async";
+import nacl from "tweetnacl";
+import naclutil from "tweetnacl-util";
+
+import { isLedgerKey } from "../KeyHelpers";
+
+import {
+  EncryptedKey,
+  EncryptedLedgerKey,
+  EncryptedPlaintextKey,
+  Key,
+  PlaintextKey,
+} from "../types";
+
+const NAME = "ScryptEncrypter";
+
+export const RECOVERY_CODE_NBITS = 160;
+export const RECOVERY_CODE_NWORDS = (RECOVERY_CODE_NBITS / 32) * 3;
+export const SALT_BYTES = 32;
+export const NONCE_BYTES = nacl.secretbox.nonceLength; // 24 bytes
+export const LOCAL_KEY_BYTES = nacl.secretbox.keyLength; // 32 bytes
+export const CRYPTO_V1 = 1;
+export const CURRENT_CRYPTO_VERSION = CRYPTO_V1;
+export const KEY_LEN = nacl.secretbox.keyLength; // 32 bytes
+
+/**
+ * Convert password from user into a derived key for encryption
+ * @param {string} param.password plaintext password from user
+ * @param {string} param.salt salt (should be randomly generated)
+ * @param {number} param.dkLen length of the derived key to output
+ * @returns {Uint8Array} bytes of the derived key
+ */
+function scryptPass({
+  password,
+  salt,
+  dkLen = KEY_LEN,
+}: {
+  password: string;
+  salt: string;
+  dkLen?: number;
+}): Promise<Uint8Array> {
+  const [N, r, p] = [32768, 8, 1];
+  return new Promise((resolve, reject) => {
+    scrypt(
+      password,
+      salt,
+      { N, r, p, dkLen, encoding: "binary" },
+      (derivedKey: Uint8Array) => {
+        if (derivedKey) {
+          resolve(derivedKey);
+        } else {
+          reject(new Error("scryptPass failed, derivedKey is null"));
+        }
+      },
+    );
+  });
+}
+
+function generateSalt(): string {
+  return naclutil.encodeBase64(nacl.randomBytes(SALT_BYTES));
+}
+
+/**
+ * "Encrypt" keys in a very basic, naive way.
+ */
+export const ScryptEncrypter = {
+  name: NAME,
+  async encryptKey({
+    key,
+    password,
+  }: {
+    key: Key;
+    password: string;
+  }): Promise<EncryptedKey> {
+    const salt = generateSalt();
+
+    if (isLedgerKey(key)) {
```

Comment:
I think we should avoid this kind of check, where an encrypter switches on the type of a key. We want encryptors to know nothing about different key types. In general, none of these should know anything about how it's being used with any of the others:

- key store
- encrypter
- key type handler

They should only interact via their publicly exposed APIs.

So in this specific scenario, the ledger should set the `privateKey` to be `null` or `""`. The encrypter / decrypter can then check for that case and skip encryption for the empty string.

The reason for this is that later, we may want to add a Trezor key handler. It should *just work* with existing encrypters and keystores, without having to update them to include special-case checks.